From 96536d645c20beb058dee7e04cb6a434d4d62975 Mon Sep 17 00:00:00 2001
From: Dimitry Ishenko <dimitry.ishenko@gmail.com>
Date: Sun, 6 Jul 2025 18:02:14 -0400
Subject: [PATCH] feat(termdebug): allow custom signs
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

For example, one can do something like this:

```vim
let g:termdebug_config.signs = ["❶", "➋", "➌", "➍", "➎", "➏", "➐", "➑", "➒"]
let g:termdebug_config.sign = "⚫"
```

where the first 9 breakpoint signs will have numbers, and the rest will
be just black circles.
---
 runtime/doc/terminal.txt                             | 2 ++
 runtime/pack/dist/opt/termdebug/plugin/termdebug.vim | 4 +++-
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/runtime/doc/terminal.txt b/runtime/doc/terminal.txt
index 36ae8fa735..7fcb89fd92 100644
--- a/runtime/doc/terminal.txt
+++ b/runtime/doc/terminal.txt
@@ -657,6 +657,8 @@ than 99 will be displayed as "9+".
 
 If you want to customize the breakpoint signs to show `>>` in the signcolumn: >vim
 	let g:termdebug_config['sign'] = '>>'
+You can also specify individual signs for the first N breakpoints: >vim
+	let g:termdebug_config['signs'] = ['a', 'b', 'c', 'x', 'y', 'z']
 If you would like to use decimal (base 10) breakpoint signs: >vim
 	let g:termdebug_config['sign_decimal'] = 1
 If the variable g:termdebug_config does not yet exist, you can use: >vim
diff --git a/runtime/pack/dist/opt/termdebug/plugin/termdebug.vim b/runtime/pack/dist/opt/termdebug/plugin/termdebug.vim
index b6da91a833..9e65daddd4 100644
--- a/runtime/pack/dist/opt/termdebug/plugin/termdebug.vim
+++ b/runtime/pack/dist/opt/termdebug/plugin/termdebug.vim
@@ -1688,7 +1688,9 @@ func s:CreateBreakpoint(id, subid, enabled)
       let hiName = "debugBreakpoint"
     endif
     let label = ''
-    if exists('g:termdebug_config') && has_key(g:termdebug_config, 'sign')
+    if exists('g:termdebug_config') && has_key(g:termdebug_config, 'signs')
+      let label = get(g:termdebug_config.signs, a:id - 1, get(g:termdebug_config, 'sign', ''))
+    elseif exists('g:termdebug_config') && has_key(g:termdebug_config, 'sign')
       let label = g:termdebug_config['sign']
     elseif exists('g:termdebug_config') && has_key(g:termdebug_config, 'sign_decimal')
       let label = printf('%02d', a:id)
-- 
2.43.0

